File: examples/multilanguage/go/aspect_fit.go
// Go + Ebiten: return scale and offset for drawing a 960x720 logical image.
package aspectfit

import "math"

const LogicalWidth = 960
const LogicalHeight = 720

type Viewport struct { X, Y, Width, Height int; Scale float64 }

func Calculate(screenWidth, screenHeight int) Viewport {
    scale := math.Min(float64(screenWidth)/LogicalWidth, float64(screenHeight)/LogicalHeight)
    width := int(math.Round(LogicalWidth * scale))
    height := int(math.Round(LogicalHeight * scale))
    return Viewport{X: (screenWidth - width) / 2, Y: (screenHeight - height) / 2, Width: width, Height: height, Scale: scale}
}

File: examples/multilanguage/csharp/AspectFit.cs
// C# + MonoGame: use this destination rectangle when drawing a 960x720 RenderTarget2D.
using Microsoft.Xna.Framework;

public static class AspectFit {
    public const int LogicalWidth = 960;
    public const int LogicalHeight = 720;

    public static Rectangle Viewport(int backBufferWidth, int backBufferHeight) {
        float scale = System.MathF.Min(backBufferWidth / (float)LogicalWidth,
                                       backBufferHeight / (float)LogicalHeight);
        int width = (int)(LogicalWidth * scale + 0.5f);
        int height = (int)(LogicalHeight * scale + 0.5f);
        return new Rectangle((backBufferWidth - width) / 2, (backBufferHeight - height) / 2, width, height);
    }
}
